src/rez/plugin_managers.py

```diff
@@ -355,10 +359,22 @@ def get_plugins(self, plugin_type):
         type."""
         return self._get_plugin_type(plugin_type).plugin_classes.keys()
 
-    def get_plugin_class(self, plugin_type, plugin_name):
+    @overload
+    def get_plugin_class(self, plugin_type: str, plugin_name: str) -> type:
+        pass
+
+    @overload
+    def get_plugin_class(self, plugin_type: str, plugin_name: str, expected_type: type[T]) -> type[T]:
+        pass
+
+    def get_plugin_class(self, plugin_type, plugin_name, expected_type=None):
         """Return the class registered under the given plugin name."""
         plugin = self._get_plugin_type(plugin_type)
```

Comment:
Added a new argument here to validate the returned result.   This provides both runtime and static validation.

Reply:
this uses type vars to describe a [generic function](https://mypy.readthedocs.io/en/stable/generics.html#generic-functions), i.e. a function where the types of its arguments are related to each other.  In our case, the type of the argument `expected_type` is related to the output type. 

Using a literal would mean we would have to define a literal value and a function overload for every possible output type, like this: 

```python
    @overload
    def get_plugin_class(self, plugin_type: str, plugin_name: str, expected_type: Literal["Foo"]) -> type[Foo]:
        pass

    @overload
    def get_plugin_class(self, plugin_type: str, plugin_name: str, expected_type: Literal["Bar"]) -> type[Bar]:
        pass
```

In a plugin environment where users can define their own types, we obviously cannot define a string constant for every possible type.

Reply:
Users can't define their own types. They can only create new plugin of some pre-defined types. Anybody adding new plugin types should do so in rez itself and not outside.